plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs

```diff
@@ -0,0 +1,251 @@
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocModified _uri
+        addFileOfInterest ide file Modified{firstOpen=False}
+        restartCabalShakeSession ide vfs file "(changed)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidSave $
+      \ide vfs _ (DidSaveTextDocumentParams TextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocSaved _uri
+        addFileOfInterest ide file OnDisk
+        restartCabalShakeSession ide vfs file "(saved)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidClose $
+      \ide vfs _ (DidCloseTextDocumentParams TextDocumentIdentifier{_uri}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocClosed _uri
+        deleteFileOfInterest ide file
+        restartCabalShakeSession ide vfs file "(closed)"
+  ]
+  }
+  where
+    log' = logWith recorder
+
+    whenUriFile :: Uri -> (NormalizedFilePath -> IO ()) -> IO ()
+    whenUriFile uri act = whenJust (LSP.uriToFilePath uri) $ act . toNormalizedFilePath'
+
+-- | Helper function to restart the shake session, specifically for modifying .cabal files.
+-- No special logic, just group up a bunch of functions you need for the base
+-- Notification Handlers.
+restartCabalShakeSession :: IdeState -> VFS.VFS -> NormalizedFilePath -> String -> IO ()
+restartCabalShakeSession ide vfs file actionMsg = do
+  join $ atomically $ Shake.recordDirtyKeys (shakeExtras ide) GetModificationTime [file]
+  restartShakeSession (shakeExtras ide) (VFSModified vfs) (fromNormalizedFilePath file ++ " " ++ actionMsg) []
+
+-- ----------------------------------------------------------------
+-- Plugin Rules
+-- ----------------------------------------------------------------
+
+data ParseCabal = ParseCabal
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable ParseCabal
+instance NFData   ParseCabal
+
+type instance RuleResult ParseCabal = ()
+
+cabalRules :: Recorder (WithPriority Log) -> Rules ()
+cabalRules recorder = do
+  -- Make sure we initialise the cabal files-of-interest.
+  ofInterestRules recorder
+  -- Rule to produce diagnostics for cabal files.
+  define (cmapWithPrio LogShake recorder) $ \ParseCabal file -> do
+    t <- use GetModificationTime file
+    log' Debug $ LogModificationTime file t
+    mVirtualFile <- Shake.getVirtualFile file
+    contents <- case mVirtualFile of
+      Just vfile -> pure $ Encoding.encodeUtf8 $ VFS.virtualFileText vfile
+      Nothing -> do
+        liftIO $ BS.readFile $ fromNormalizedFilePath file
+
+    (pWarnings, pm) <- liftIO $ Parse.parseCabalFileContents contents
+    let warningDiags = fmap (Diagnostics.warningDiagnostic file) pWarnings
+    case pm of
+      Left (_cabalVersion, pErrorNE) -> do
+        let errorDiags = NE.toList $ NE.map (Diagnostics.errorDiagnostic file) pErrorNE
+            allDiags = errorDiags <> warningDiags
+        log' Debug $ LogDiagnostics file allDiags
```

Comment:
This logging seems very generic. Maybe we should put it in the shake code so it gets logged for every rule that produces diagnostics?